python_modules/dagster-ext/dagster_ext_tests/test_context.py

```diff
@@ -162,3 +162,10 @@ def test_extras_context():
     assert context.get_extra("foo") == "bar"
     with pytest.raises(DagsterExtError, match="Extra `bar` is undefined"):
         context.get_extra("bar")
+
+
+def test_report_after_materialization():
+    context = _make_external_execution_context(asset_keys=["foo"])
+    with pytest.raises(DagsterExtError, match="already been materialized"):
+        context.report_asset_materialization(asset_key="foo")
+        context.report_asset_data_version("alpha", asset_key="foo")
```

Comment:
I don't think this is an acceptable DX and therefore we should change the context API to make this error impossible

Reply:
Does `report_asset_data_version` update an existing asset's data version if that asset is not being materialized in this step?

Reply:
>change the context API to make this error impossible

Should we just get rid of the piecewise reporting methods (`report_asset_data_version`, `report_asset_metadata`) and require the user to report all info at once via `report_asset_materialization`? While working on this I was thinking that would be better but didn't want to completely upend the apple cart before review.

Reply:
> Should we just get rid of the piecewise reporting methods (report_asset_data_version, report_asset_metadata) and require the user to report all info at once via report_asset_materialization? While working on this I was thinking that would be better but didn't want to completely upend the apple cart before review.

Yes let's remove this

Reply:
Do it in a sep PR and stack this on top

Reply:
>Do it in a sep PR and stack this on top

Not sure how to do this-- in `master`, `report_asset_metadata` and `report_asset_data_version` are implemented by calling corresponding `OpExecutionContext` methods, which handle attaching the payload to the ultimately generated contexts. AFAICT there is no corresponding method that would allow stashing a `MaterializeResult` on the `OpExecutionContext`, so a separate temporary `OpExecutionContext` API would need to be added just for this downstack PR. We can remove them in the present PR because we are yielding/returning `MaterializeResult`.